pkg/apiserver/apiserver.go

```diff
@@ -198,31 +201,42 @@ func newServerConfig(aggClient aggregatorclient.Interface, opts NewAPIServerOpts
 	if err != nil {
 		return nil, err
 	}
-	// this feature gate is not enabled in k8s <1.26 as the
-	// APIs it relies on were in v1beta2/v1beta1/alpha.
-	// the apiserver library hardcodes the v1beta3 version of the resource
+
+	// this feature gate is not enabled in k8s <1.29 as the
+	// APIs it relies on were in v1beta3/v1beta2/v1beta1/alpha.
+	// the apiserver library hardcodes the v1 version of the resource
 	// so the best we can do for older k8s clusters is to allow it to be disabled.
-	minSupportedVersionForAPF, err := semver.New("1.26.0")
+	minSupportedVersionForAPF, err := semver.New("1.29.0")
 	if err != nil {
 		return nil, err
 	}
 	isServerVerLTminSupportedVer := serverVersion.LT(*minSupportedVersionForAPF)
 	if !opts.EnableAPIPriorityAndFairness || isServerVerLTminSupportedVer {
 		if isServerVerLTminSupportedVer {
-			opts.Logger.Info("The current version of kapp-controller does not support api-priority-and-fairness for versions of kubernets prior to 1.26, disabling this option")
-		}
-		err := feature.DefaultMutableFeatureGate.Set("APIPriorityAndFairness=false")
-		if err != nil {
-			return nil, fmt.Errorf("error updating disabling feature gate for APIPriorityAndFairness: %v", err)
+			opts.Logger.Info("The current version of kapp-controller does not support api-priority-and-fairness for versions of kubernets prior to 1.29, disabling this option")
 		}
+		recommendedOptions.Features.EnablePriorityAndFairness = false
+	}
+
+	// validatingAdmissionPolicy has been made available by default for K8s >= 1.30.
+	// We will remove the validatingAdmissionPolicy from the execution in case K8s < 1.30.
+	// However, we will still run namespaceLifecycle, mutatingAdmissionWebhook and validatingAdmissionWebhooks.
+	minSupportedVersionForValidatingAdmissionPolicy, err := semver.New("1.30.0")
+	if err != nil {
+		return nil, err
+	}
+	isServerVerLTminSupportedVer = serverVersion.LT(*minSupportedVersionForValidatingAdmissionPolicy)
+	if isServerVerLTminSupportedVer {
+		recommendedOptions.Admission.RecommendedPluginOrder = []string{lifecycle.PluginName, mutatingwebhook.PluginName, validatingwebhook.PluginName}
 	}
 
 	serverConfig := genericapiserver.NewRecommendedConfig(Codecs)
+	recommendedOptions.Features.EnablePriorityAndFairness = false
```

Comment:
This would always set APF to false, right?

Reply:
Yeah, looks like some rough work creeped in. Will remove it.